### PR TITLE
Add global optimization

### LINKED ISSUE #96: Global optimization with outlier detection
Supersedes #44.

### LINKED ISSUE #109: Keep offsets instead of transforms internally
in `m_TransformCandidates` and `m_CurrentAdjustments`

### PR DESCRIPTION
**Pull request by dzenanz**
Closes #96. Closes #109.

### CODE REVIEW

File: include/itkTileMontage.h [hunk 1]
```diff
@@ -119,6 +119,18 @@ class ITK_TEMPLATE_EXPORT TileMontage : public ProcessObject
   itkSetMacro( ForcedSpacing, SpacingType );
   itkGetConstMacro( ForcedSpacing, SpacingType );
 
+  /** Set/Get absolute registration threshold.
+   * The maximum allowed residual error for a registration pair during
+   * global optimization. Expressed in number of pixels. Default: 1.0. */
```
Comment (thewtex) [inline include/itkTileMontage.h]: Please make a note here about what happens when a registration pair's error exceeds the threshold.

File: include/itkTileMontage.h [hunk 2]
```diff
@@ -119,6 +119,18 @@ class ITK_TEMPLATE_EXPORT TileMontage : public ProcessObject
...
+  itkSetMacro( AbsoluteThreshold, float );
+  itkGetConstMacro( AbsoluteThreshold, float );
+
+  /** Set/Get relative registration threshold.
+   * The maximum allowed deviation for a registration pair during global
+   * optimization. Expressed in multiples of standard deviation. Default: 3.0. */
```
Comment (thewtex) [inline include/itkTileMontage.h]: Please make a note here indicating how the standard deviation is computed.

File: include/itkTileMontage.hxx [hunk 3]
```diff
@@ -24,8 +24,12 @@
 #include "itkMultiThreaderBase.h"
 #include "itkNumericTraits.h"
 
+#include "itk_eigen.h"
+#include ITK_EIGEN( Sparse )
```
Comment (thewtex) [inline include/itkTileMontage.hxx]: Just curious -- what does this do?

File: include/itkTileMontage.hxx [hunk 4]
```diff
@@ -472,7 +449,213 @@ void
...
+      const TranslationOffset& candidateOffset = m_TransformCandidates[i][0];
+      for ( unsigned d = 0; d < ImageDimension; d++ )
+        {
+        translations( regIndex, d ) = candidateOffset[d];
+        }
+      equationToCandidate[regIndex] = i;
+      ++regIndex;
+      }
+    }
+  std::cout << std::endl;
```
Comment (thewtex) [inline include/itkTileMontage.hxx]: Remove debug code

File: include/itkTileMontage.hxx [hunk 5]
```diff
@@ -472,7 +449,213 @@ void
...
+
+      if ( !m_TransformCandidates[candidateIndex].empty() )
+        {
+        // get a new equation from m_TransformCandidates
+        const TranslationOffset& candidateOffset = m_TransformCandidates[candidateIndex][0];
+        for ( unsigned d = 0; d < ImageDimension; d++ )
+          {
+          translations( maxIndex, d ) = candidateOffset[d];
+          }
+        std::cout << "  Replaced by T: " << candidateOffset;
```
Comment (thewtex) [inline include/itkTileMontage.hxx]: Remove debug code

File: include/itkTileMontage.hxx [hunk 6]
```diff
@@ -472,7 +449,213 @@ void
...
+        typename SparseMatrix::InnerIterator it( regCoef, maxIndex );
+        regCoef.coeffRef( maxIndex, it.index() ) = 0.01 * regCoef.coeffRef( maxIndex, it.index() );
+        ++it;
+        regCoef.coeffRef( maxIndex, it.index() ) = 0.01 * regCoef.coeffRef( maxIndex, it.index() );
+
+        for ( unsigned d = 0; d < ImageDimension; d++ )
+          {
+          translations( maxIndex, d ) = 0;
+          }
+        std::cout << "  Replaced by zeroes.";
```
Comment (thewtex) [inline include/itkTileMontage.hxx]: Remove debug code

File: include/itkTileMontage.hxx (hunk 3 above)
Comment (dzenanz) [reply include/itkTileMontage.hxx]: Selects either internal Eigen's header, or system Eigen's header, based on CMake configuration. I copy-pasted this snippet from SymmetricEigenAnalysis class which @phcerdan modified after adding Eigen to ITK.

File: include/itkTileMontage.hxx (hunk 5 above)
Comment (dzenanz) [reply include/itkTileMontage.hxx]: This is more of a legitimate info message, than debug code.

File: include/itkTileMontage.hxx (hunk 6 above)
Comment (dzenanz) [reply include/itkTileMontage.hxx]: Informative message, not debug code. This finishes the sentence started by `Outlier detected. Equation ...`.

**Pull request merged**